File: Week_03/kotlin/src/_60_PermutationSequence.kt
fun getPermutation(n: Int, k_: Int): String {
    var k = k_
    val factorial = IntArray(n).apply { this[0] = 1 }
    for (i in 1 until n) {
        factorial[i] = factorial[i - 1] * i
    }
    k--
    val ret = StringBuilder()
    val valid = IntArray(n + 1) { 1 }
    for (i in 1..n) {
        var order = k / factorial[n - i] + 1;
        for (j in 1..n) {
            order -= valid[j];
            if (order == 0) {
                ret.append(j);
                valid[j] = 0;
                break;
            }
        }
        k %= factorial[n - i];
    }
    return ret.toString()
}

File: Week_03/cpp/src/_60_PermutationSequence.cpp
#include <vector>
#include <string>

using namespace std;

class Solution
{
public:
    string getPermutation(int n, int k)
    {
        vector<int> factorial(n);
        factorial[0] = 1;
        for (auto i = 1; i < n; i++)
        {
            factorial[i] = factorial[i - 1] * i;
        }
        k--;
        string ret;
        vector<int> valid(n + 1, 1);
        for (auto i = 1; i <= n; i++)
        {
            auto order = k / factorial[n - i] + 1;
            for (auto j = 1; j <= n; j++)
            {
                order -= valid[j];
                if (!order)
                {
                    ret += (j + '0');
                    valid[j] = 0;
                    break;
                }
            }
            k %= factorial[n - i];
        }
        return ret;
    }
};